### ISSUE TITLE
Object cache control headers is wrong on stream

### ISSUE THREAD

**Issue by zlk89** [googleapis/nodejs-storage#1985]

**Comment by zlk89** [comment]
@ddelgrosso1 Good to know. What would be the recommended way to get the cache control headers without making additional request for metadata in the example?

**Comment by zlk89** [comment]
Right, we don't want to make the bucket publicly accessible. We are gating the whole bucket and proxy service for internal users only.

One solution would be making a call to `getMetadata` as you mentioned:
```
app.get("*", async (req, res, next) => {
    const file = bucket.file(req.path);
    const metadata = await file.getMetadata(); // <- extra call?
    // set headers with metadata here
    file.createReadStream().pipe(res);
  });
```
But this seems to make two calls for one proxy request.

Is there a better way to handle it?

**Comment by ddelgrosso1** [comment]
You can avoid the additional call to `getMetadata` by hooking into the `.on('end')` event of the read stream. At that point in the cycle `file.metadata` should be populated.

**Comment by zlk89** [comment]
Unfortunately, I am getting an Error `Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client` when I set the headers in `.on('end')`...

**Comment by ddelgrosso1** [comment]
Apologies, I misread the code and what you were attempting to do. I believe in this situation you are correct, if you wish to set the express response headers and pipe the file data to the response it will necessitate an additional call to `getMetadata`. The metadata is available after the stream is complete but as you noted you cannot set the headers at that point as the stream has already been piped to the response. The only other thing I can think would be to buffer the file data in memory before sending it to the express response which may not be feasible depending on file size / system load.

**Comment by ddelgrosso1** [comment]
I believe that the original question regarding cache-control headers was answered. I am going to close this issue as a result. If you feel that there are still questions or other issues outstanding. Please feel free to reopen.
